[SYCL][HIP] Port HIP plugin to Unified Runtime

omarahmed1111 opened the pull request:
This moves the HIP plugin implementation to Unified Runtime; and changes the pi_hip plugin to use pi2ur to implement PI. The changes to the implementation have been kept to a minimum and should be functionally the same. Documentation and comments have been moved verbatim, other than changing PI references to UR.

This PR is based on top of the CUDA adapter (https://github.com/intel/llvm/pull/9512) so will only be ready when that is merged.

jandres742 commented on sycl/plugins/hip/CMakeLists.txt:
```diff
@@ -87,12 +87,43 @@ set(HIP_HEADERS "${PI_HIP_INCLUDE_DIR};${PI_HIP_HSA_INCLUDE_DIR}")
 # Create pi_hip library
 add_sycl_plugin(hip
   SOURCES
+    # Some code is shared with the UR adapter
```
consider adding versioning as we have for CUDA and L0.

jandres742 commented:
looks good.

dm-vodopyanov commented:
@omarahmed1111 please don't do force-push - it complicates code review process for reviewers as the information about what changed from commit to commit is lost.

omarahmed1111 commented:
> @omarahmed1111 please don't do force-push - it complicates code review process for reviewers as the information about what changed from commit to commit is lost.

My apologies, I thought the process of reviewing didn't start yet, will gradually change it next time with pushing new commits.

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/context.cpp:
```diff
@@ -0,0 +1,174 @@
+  case UR_CONTEXT_INFO_ATOMIC_FENCE_SCOPE_CAPABILITIES: {
+    // These queries should be dealt with in context_impl.cpp by calling the
+    // queries of each device separately and building the intersection set.
+    setErrorMessage("These queries should have never come here.",
```
Once you merge this the error message set here will be lost due to: https://github.com/intel/llvm/issues/10066 
I suggest switching to UR_CHECK_ERROR usage temporarily so that the error message is preserved: like this change: https://github.com/intel/llvm/pull/8303/commits/37aecb11e768fdb4ad3c590d83b3ac4f5e03db7b

omarahmed1111 replied:
Actually, I think this should be `UR_RESULT_ERROR_INVALID_ENUMERATION` rather than `UR_RESULT_ERROR_ADAPTER_SPECIFIC` as the error comes from that this queries shouldn't be dealt with in UR (as the comment explains). So, I think it is more of an error from UR layer than CUDA layer, does that sound reasonable?

omarahmed1111 replied:
or also could make sense to be `ur::assertion` if we should not return gracefully in this case.

omarahmed1111 replied:
I have taken the path of making this return `UR_RESULT_ERROR_INVALID_ENUMERATION` as this should be a UR layer error, and the comment was explaining what happens enough to not support that.

JackAKirk commented:
Did you read through the review of https://github.com/intel/llvm/pull/9512 ?
Since hip PI copied cuda PI, a lot of the comments there may be relevant here.

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/queue.cpp:
```diff
@@ -0,0 +1,280 @@
+  std::ignore = hDevice;
+  std::ignore = pProperties;
+  std::ignore = phQueue;
+  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
```
It seems like doing this is becoming the standard way of dealing with this https://github.com/intel/llvm/pull/8303/commits/f2a2026c16de51b68de60b503414f93aa43fb699#diff-8da59f0730e70207732121ceff2865ddb20a577d248e7cf4709658de4fc8a579R19 (I copied from other PRs)
Currently I don't think the programmer will get any error message currently otherwise.

omarahmed1111 replied:
We were trying to avoid calling `die` as much as possible as that would terminate the program, I would suggest that we could resolve the `die` call in another PR as that might prove controversial and is connected to the way that adapters specific error handling would be done. Also, I would prefer that the port here not change any behaviour that was there as that could introduce other unexpected problems.

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/program.cpp:
```diff
@@ -0,0 +1,320 @@
+  }
+
+  return Result;
+}
```
missing EOF

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/device.cpp:
```diff
@@ -0,0 +1,999 @@
+    return ReturnValue(Value);
+  }
+  case UR_DEVICE_INFO_UUID: {
+#if ((HIP_VERSION_MAJOR == 5 && HIP_VERSION_MINOR >= 2) ||                     \
```
This is strange, the comment below says it is supported since 5.2, then the logic to the check only requires >5.0, but then does a check for 5.2 first. Either this is a bug and it should only be checking for 5.2 or the comment is wrong and it should only be checking for 5.0 if it is supported since 5.0 rather than 5.2.

hdelan replied:
I think this is OK. The first check allows 5.3 say to pass, and the next check allows 6.1 to pass since only maj version is being checked

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/device.cpp:
```diff
@@ -0,0 +1,999 @@
+  }
+
+  case UR_DEVICE_INFO_ATOMIC_64: {
+    // TODO: Reconsider it when AMD supports SYCL_USE_NATIVE_FP_ATOMICS.
```
I think this comment is out of date since native atomics now supported? @hdelan

hdelan replied:
Yep comment can be removed. Logic is OK though

omarahmed1111 replied:
Fixed

hdelan commented:
> > @omarahmed1111 please don't do force-push - it complicates code review process for reviewers as the information about what changed from commit to commit is lost.
> 
> My apologies,
> 
> I only won't be able to rebase without force-pushing. Could you explain to me so I would have a better idea of how to resolve that? @dm-vodopyanov

I think usually they recommend to have a merge commit instead of a rebase

omarahmed1111 commented:
> Did you read through the review of #9512 ? Since hip PI copied cuda PI, a lot of the comments there may be relevant here.

Yes, I addressed all the comments that was addressed in CUDA porting, however, I skipped the ones that was not resolved in there as I assumed that this didn't resolve to a decision regarding it, like this [one](https://github.com/intel/llvm/pull/9512#discussion_r1218096906)

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/usm.cpp:
```diff
@@ -0,0 +1,212 @@
+}
+
+/// USM: Implements USM device allocations using a normal HIP device pointer
+UR_APIEXPORT ur_result_t UR_APICALL urUSMDeviceAlloc(
```
Addressed

jchlanda commented:
> > @omarahmed1111 please don't do force-push - it complicates code review process for reviewers as the information about what changed from commit to commit is lost.
> 
> My apologies,
> 
> I only won't be able to rebase without force-pushing. Could you explain to me so I would have a better idea of how to resolve that? @dm-vodopyanov

Once the PR is opened, the dominant workflow is to merge in `sycl` into your development branch (which would avoid the need for force pushing), for some background discussion see: https://github.com/intel/llvm/discussions/8229

omarahmed1111 commented:
> I think usually they recommend to have a merge commit instead of a rebase

> Once the PR is opened, the dominant workflow is to merge in `sycl` into your development branch (which would avoid the need for force pushing), for some background discussion see: #8229

That make sense, thanks for clarification @hdelan @jchlanda

omarahmed1111 commented:
@intel/llvm-reviewers-runtime Just wanted to ping on reviewing the PR, Thanks!

ldrumm commented on sycl/plugins/unified_runtime/ur/adapters/hip/kernel.hpp:
```diff
@@ -0,0 +1,194 @@
+  /// This is not something can be queried from the HIP API
+  /// so there is a hard-coded size (\ref MAX_PARAM_BYTES)
+  /// and a storage.
+  ///
```
```suggestion
```

ldrumm commented on sycl/plugins/unified_runtime/ur/adapters/hip/program.cpp:
```diff
@@ -0,0 +1,320 @@
+                      size_t length, const ur_program_properties_t *pProperties,
+                      ur_program_handle_t *phProgram) {
+  ur_device_handle_t hDevice = hContext->getDevice();
+  auto pBinary = reinterpret_cast<const uint8_t *>(pIL);
```
const

ldrumm commented on sycl/plugins/unified_runtime/ur/adapters/hip/program.cpp:
```diff
@@ -0,0 +1,320 @@
+/// HIP will handle the PTX/HIPBIN binaries internally through a call to
+/// hipModuleLoadDataEx. So, urProgramCompile and urProgramBuild are equivalent
+/// in terms of HIP adapter. \TODO Implement asynchronous compilation
+///
```
```suggestion
```

ldrumm reviewed (changes requested):
Great work!

I've tested a previous iteration of this patch locally and it appears to be working well, so my review is mostly local, rather than architectural, and often simply stylistic.

Feel free to push back or ignore some of the style changes requested, but there are a couple of potential bugs that do need addressing

ldrumm reviewed (commented):
I seem to have managed to split my review in two somehow; apologies

ldrumm commented on sycl/plugins/unified_runtime/ur/adapters/hip/event.cpp:
```diff
@@ -0,0 +1,327 @@
+}
+
+uint64_t ur_event_handle_t_::getStartTime() const {
+  float MiliSeconds = 0.0f;
+  assert(isStarted());
+
+  // hipEventSynchronize waits till the event is ready for call to
+  // hipEventElapsedTime.
+  UR_CHECK_ERROR(hipEventSynchronize(ur_platform_handle_t_::EvBase));
+  UR_CHECK_ERROR(hipEventSynchronize(EvStart));
+
+  UR_CHECK_ERROR(hipEventElapsedTime(&MiliSeconds,
+                                     ur_platform_handle_t_::EvBase, EvStart));
+  return static_cast<uint64_t>(MiliSeconds * 1.0e6);
```
Oh I get it. You want to return nanoseconds. Silly me!

omarahmed1111 commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -172,6 +172,64 @@ if ("cuda" IN_LIST SYCL_ENABLE_PLUGINS)
+  )
+
+  set_target_properties("ur_adapter_hip" PROPERTIES
+    VERSION "0.0.0"
```
We would expect that to have the UR version, but atm, we align that with other adapters which have a value of `0`, and we would bump all of the adapters in a seperate PR so they align.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.hpp:
```diff
@@ -0,0 +1,176 @@
+#endif
+}
+
+// NVidia HIP headers guard hipArray3DCreate behind __CUDACC__, this does not
```
It was ported from the pi hip plugin [here](https://github.com/intel/llvm/blob/f1986b40e04b7fa0bfeb5bc95ad2212376522528/sycl/plugins/hip/pi_hip.cpp#L51). It is related to hip on top of cuda build.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.hpp:
```diff
@@ -0,0 +1,176 @@
+// NVidia HIP headers guard hipArray3DCreate behind __CUDACC__, this does not
+// seem to be required and we're not using nvcc to build the UR HIP adapter so
+// add the translation function here
+#if defined(__HIP_PLATFORM_NVIDIA__) && !defined(__CUDACC__)
```
Not sure, but I don't think so, I think the comment mentions that this issue happens because we are not using nvcc but doesn't imply that it did exist.

npmiller commented on sycl/plugins/unified_runtime/ur/adapters/hip/usm.cpp:
```diff
@@ -0,0 +1,197 @@
+  }
+  UR_ASSERT(!pUSMDesc || (pUSMDesc->align == 0 ||
+                          ((pUSMDesc->align & (pUSMDesc->align - 1)) == 0)),
+            UR_RESULT_ERROR_INVALID_VALUE);
```
This check is already done above

omarahmed1111 replied:
Fixed

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.hpp:
```diff
@@ -0,0 +1,176 @@
+extern thread_local char ErrorMessage[MaxMessageSize];
+
+// Utility function for setting a message and warning
+[[maybe_unused]] void setErrorMessage(const char *Message,
```
This is to suppress the warning on using the function.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/context.cpp:
```diff
@@ -0,0 +1,172 @@
+  std::ignore = DeviceCount;
+  std::ignore = pProperties;
+
+  assert(DeviceCount == 1);
```
It just to enforce the usage of only using one device in sycl, in the future we may support multiple devices, but this is just to make sure that only one device is used atm, as multiple devices are still not supported.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.hpp:
```diff
@@ -0,0 +1,176 @@
+/// object on destruction.
+///
+/// The `dismiss` function stops the release from happening on destruction.
+template <typename T> class ReleaseGuard {
```
It will already be enforced by the compiler as if we tried to use any type other than pointer type, the compilation will fail in this [lines](https://github.com/intel/llvm/blob/663042b04b63cb1830ae12b4442e12af4830764e/sycl/plugins/hip/pi_hip.cpp#L792) as they will give an invalid conversion, I think `enable_if` could give more clarity to this enforcement but I feel that it may also complicate this class than necessary.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/device.cpp:
```diff
@@ -0,0 +1,996 @@
+    return_sizes.sizes[0] = size_t(MaxX);
+    return_sizes.sizes[1] = size_t(MaxY);
+    return_sizes.sizes[2] = size_t(MaxZ);
+    return ReturnValue(return_sizes);
```
It will deal with it as an array, but this struct will be handled internally in UR in this [`urReturnHelper` class](https://github.com/intel/llvm/blob/663042b04b63cb1830ae12b4442e12af4830764e/sycl/plugins/unified_runtime/ur/ur.hpp#L250) which will be casted to an array.

ldrumm replied:
Gotcha.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.cpp:
```diff
@@ -0,0 +1,104 @@
+  throw mapErrorUR(Result);
+}
+
+std::string getHipVersionString() {
```
Yeah, good point, fixed that.

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/event.cpp:
```diff
@@ -0,0 +1,327 @@
+  UR_CHECK_ERROR(hipEventSynchronize(EvEnd));
+
+  UR_CHECK_ERROR(hipEventElapsedTime(&MiliSeconds, EvStart, EvEnd));
+  return static_cast<uint64_t>(MiliSeconds * 1.0e6);
```
I fixed the typo to `MilliSecond`, as I felt `MS` could lead to some confusions so more clarity is better here .

omarahmed1111 commented on sycl/plugins/unified_runtime/ur/adapters/hip/common.cpp:
```diff
@@ -0,0 +1,104 @@
+       << "\n\tFunction:        " << Function << "\n\tSource Location: " << File
+       << ":" << Line << "\n"
+       << std::endl;
+    std::cerr << SS.str();
```
fixed that and deleted the `stringstream`

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/command_buffer.cpp:
```diff
@@ -0,0 +1,129 @@
+UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferCreateExp(
+    ur_context_handle_t, ur_device_handle_t,
+    const ur_exp_command_buffer_desc_t *, ur_exp_command_buffer_handle_t *) {
+  detail::ur::die("Experimental Command-buffer feature is not "
```
I was told to use this instead of `ur::die`: https://github.com/intel/llvm/pull/8303/files#r1253525859
Would be good to be consistent on this.

omarahmed1111 replied:
`L0` uses a different structure to report errors than `CUDA/HIP` atm. `urPrint` depends on env variables that are defined inside `L0` plugin only atm like `UR_L0_DEBUG`, so to port this functionality across to CUDA/HIP, We will need to generalize this env variables and uses them across all the adapters. So, this is just a temporary solution so that cuda and hip align with each other. And, I guess we could work on generalizing `urPrint` in a different PR as it could be out of context in here to work on redefining those env variables across adapters.

JackAKirk replied:
OK I see thanks.

JackAKirk commented on sycl/plugins/unified_runtime/ur/adapters/hip/usm_p2p.cpp:
```diff
@@ -0,0 +1,31 @@
+UR_APIEXPORT ur_result_t UR_APICALL
+urUsmP2PPeerAccessGetInfoExp(ur_device_handle_t, ur_device_handle_t,
+                             ur_exp_peer_info_t, size_t, void *, size_t *) {
+  detail::ur::die(
```
This isn't the biggest deal, but this is changing the current behaviour which was to return false for all queries. You could make it do the same thing as it does in the level_zero adapter for instance: https://github.com/intel/llvm/blob/9e6f7962876c027e801608d43fd25f88ac3f81d6/sycl/plugins/unified_runtime/ur/adapters/level_zero/usm_p2p.cpp#L42

omarahmed1111 replied:
Changed that.

jandres742 commented:
> @jandres742 I think we might only need your approval on this PR before its ready to merge.

Looks good on my side. another adapter :), great!

omarahmed1111 commented:
@intel/llvm-gatekeepers could this be merged when possible.

kbenzie commented:
> @intel/llvm-gatekeepers could this be merged when possible.

@steffenlarsen seems like this may have slipped through the cracks yesterday, perhaps you can help us out?
